_posts/2019-07-16-domain-vertial-classifier.md

```diff
@@ -0,0 +1,179 @@
+
+The domain name dataset presents a number of other challenges; here are the the problems we dealt with when building this model:
+
+### Problems
+
+- **Problem 1**: How many categories should we keep in the model, given the difficulty in predicting multiple categories from very short text lengths?
+- **Problem 2**: How to deal with noise in the data?
+- **Problem 3**: Can we improve the model by adding network layers or increasing the embedding size?
+- **Problem 4**: Should we embed the input domain at word level or character level?
+
+
+
+## The Model
+
+To tackle the problem of classifying domain names to an industry category, we implemented a classic Convolutional Neural Network (CNN) classifier (learn more here at [Google CNN Tutorial] and [Stanford CNN Tutorial]). To generalize the model at a very high level, it takes input as a string in the form of words contained in the second-level domain (SLD) -- for example: \"www.iamanexample.com\" --> \"i am an example\", projects it into vector space by using a word embedding. Then the CNN cells of various sizes start sliding over the vector space text, to collect features in different flavors. A max pooling layer follows to get the feature mappings. Then all the locality information in the feature mappings merge through a fully connected layer for optimizing over softmax. Finally, the category predictions will be obtained from there. Please check out these excellent explanations on how a CNN and Max Pooling works: [CNN explained on Quora] and [Max Pooling explained on Quora].
+
+### Dropping Weights From A Machine Learning Model --  Less is More
+
+Now, let's talk about **Problem 1** (#problems). One important thing I learned from doing this project is we have to be realistic about technology. Sure, deep neural networks are awesome, but they are not magic. Intuitively, a deeper and more complicated network structure would be more intelligent and able to handle more profound logic. However, if the problem itself does not always have a black and white answer, there will be a cap on how well the model can perform. For example, if some newly discovered alien language is presented to you and the model, neither you nor the model would be able to decipher it, since none of you have any prior knowledge about it. Similarly, if you see just the word "apple" without any context, you'd be confused whether it refers to the tech company or the fruit, and the model will think alike. So, the answer is, we can't keep every single category. The categories shouldn't be very specific either, as there is only so much we can infer from a very short text.
```

Comment:
```suggestion
Now, let's talk about [Problem 1](#problems). One important thing I learned from doing this project is we have to be realistic about technology. Sure, deep neural networks are awesome, but they are not magic. Intuitively, a deeper and more complicated network structure would be more intelligent and able to handle more profound logic. However, if the problem itself does not always have a black and white answer, there will be a cap on how well the model can perform. For example, if some newly discovered alien language is presented to you and the model, neither you nor the model would be able to decipher it, since none of you have any prior knowledge about it. Similarly, if you see just the word "apple" without any context, you'd be confused whether it refers to the tech company or the fruit, and the model will think alike. So, the answer is, we can't keep every single category. The categories shouldn't be very specific either, as there is only so much we can infer from a very short text.
```